File: mgmt-application/src/main/java/io/quartic/mgmt/MgmtResource.java
package io.quartic.mgmt;

import com.google.common.collect.ImmutableMap;
import io.quartic.catalogue.api.CatalogueService;
import io.quartic.catalogue.api.model.CloudGeoJsonDatasetLocator;
import io.quartic.catalogue.api.model.DatasetConfig;
import io.quartic.catalogue.api.model.DatasetId;
import io.quartic.catalogue.api.model.DatasetNamespace;
import io.quartic.common.geojson.GeoJsonParser;
import io.quartic.howl.api.HowlService;
import io.quartic.howl.api.HowlStorageId;
import io.quartic.mgmt.conversion.CsvConverter;
import io.quartic.mgmt.conversion.GeoJsonConverter;
import org.apache.commons.io.IOUtils;

import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.BadRequestException;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import java.io.IOException;
import java.io.InputStream;
import java.util.Map;

import static java.util.Collections.emptyMap;

@Path("/")
public class MgmtResource {
    private static final String HOWL_NAMESPACE = "mgmt";
    private final CatalogueService catalogue;
    private final HowlService howl;
    private final DatasetNamespace defaultCatalogueNamespace;

    public MgmtResource(CatalogueService catalogue, HowlService howl, DatasetNamespace defaultCatalogueNamespace) {
        this.catalogue = catalogue;
        this.howl = howl;
        this.defaultCatalogueNamespace = defaultCatalogueNamespace;
    }

    @GET
    @Path("/dataset")
    @Produces(MediaType.APPLICATION_JSON)
    public Map<DatasetId, DatasetConfig> getDatasets() {
        return ImmutableMap.copyOf(catalogue.getDatasets().getOrDefault(defaultCatalogueNamespace, emptyMap()));
    }

    @DELETE
    @Path("/dataset/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    public void deleteDataset(@PathParam("id") DatasetId id) {
        catalogue.deleteDataset(defaultCatalogueNamespace, id);
    }

    @POST
    @Path("/dataset")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    public DatasetId createDataset(CreateDatasetRequest createDatasetRequest) {
        DatasetConfig datasetConfig = createDatasetRequest.accept(request -> {
            try {
                String name = preprocessFile(request.fileName(), request.fileType());
                return new DatasetConfig(
                        request.metadata(),
                        new CloudGeoJsonDatasetLocator(String.format("/%s/%s", HOWL_NAMESPACE, name), false),
                        emptyMap()
                );
            }
            catch (IOException e) {
                e.printStackTrace();
                throw new RuntimeException("exception while preprocessing file: " + e);
            }
        });
        return catalogue.registerDataset(defaultCatalogueNamespace, datasetConfig).getId();
    }

    private String preprocessFile(String fileName, FileType fileType) throws IOException {
        InputStream inputStream = howl.downloadFile(HOWL_NAMESPACE, fileName)
                .orElseThrow(() -> new NotFoundException("file not found: " + fileName));

        switch (fileType) {
            case GEOJSON:
                try {
                    new GeoJsonParser(inputStream).validate();
                } catch (Exception e) {
                    throw new BadRequestException("Exception while validating geojson: " + e);
                }
                return fileName;
            case CSV:
                GeoJsonConverter converter = new CsvConverter();
                HowlStorageId storageId = howl.uploadFile(MediaType.APPLICATION_JSON, HOWL_NAMESPACE,
                        outputStream -> {
                            try {
                                converter.convert(inputStream, outputStream);
                            } catch (IOException e) {
                                e.printStackTrace();
                                throw new BadRequestException("Exception while converting csv to geojson: " + e);
                            }
                        });
                return storageId.getUid();
            case RAW:
            default:
                return fileName;
        }
    }

    @POST
    @Path("/file")
    @Produces(MediaType.APPLICATION_JSON)
    public HowlStorageId uploadFile(@Context HttpServletRequest request) throws IOException {
        return howl.uploadFile(request.getContentType(), HOWL_NAMESPACE,
                outputStream -> {
                    try {
                        IOUtils.copy(request.getInputStream(), outputStream);
                    } catch (Exception e) {
                        e.printStackTrace();
                        throw new RuntimeException("exception while uploading file: " + e);
                    }
                });
    }
}

File: mgmt-application/src/main/kotlin/io/quartic/mgmt/MgmtResource.kt
package io.quartic.mgmt

import io.dropwizard.auth.Auth
import io.quartic.catalogue.api.CatalogueService
import io.quartic.catalogue.api.model.*
import io.quartic.common.auth.User
import io.quartic.common.geojson.GeoJsonParser
import io.quartic.howl.api.HowlService
import io.quartic.howl.api.HowlStorageId
import io.quartic.mgmt.auth.NamespaceAuthoriser
import io.quartic.mgmt.conversion.CsvConverter
import org.apache.commons.io.IOUtils
import java.io.IOException
import java.io.InputStream
import javax.servlet.http.HttpServletRequest
import javax.ws.rs.*
import javax.ws.rs.core.Context
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("/")
class MgmtResource(
        private val catalogue: CatalogueService,
        private val howl: HowlService,
        private val authoriser: NamespaceAuthoriser
) {
    // TODO - frontend will need to cope with DatasetNamespace in request paths and GET /datasets response

    // TODO - how does frontend know what namespace to use for dataset creation?

    @GET
    @Path("/datasets")
    @Produces(MediaType.APPLICATION_JSON)
    fun getDatasets(@Auth user: User): Response {
        return Response.ok().status(401).build()
//        catalogue.getDatasets().filterKeys { namespace -> authoriser.authorisedFor(user, namespace) }
    }


    @DELETE
    @Path("/datasets/{namespace}/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    fun deleteDataset(
            @Auth user: User,
            @PathParam("namespace") namespace: DatasetNamespace,
            @PathParam("id") id: DatasetId
    ) {
        // Note there's a potential race-condition here - another catalogue client could have manipulated the
        // dataset in-between these two statements.  It shouldn't matter - we will never delete a dataset not in an
        // authorised namespace.
        throwIfDatasetNotPresentOrNotAllowed(user, DatasetCoordinates(namespace, id))
        catalogue.deleteDataset(namespace, id)
    }

    @POST
    @Path("/datasets/{namespace}")
    @Consumes(MediaType.APPLICATION_JSON)
    @Produces(MediaType.APPLICATION_JSON)
    fun createDataset(
            @Auth user: User,
            @PathParam("namespace") namespace: DatasetNamespace,
            request: CreateDatasetRequest
    ): DatasetId {
        throwIfNamespaceNotAllowed(user, namespace)

        val datasetConfig = when (request) {
            is CreateStaticDatasetRequest -> {
                try {
                    val name = preprocessFile(namespace.namespace, request.fileName, request.fileType)
                    val locator = DatasetLocator.CloudDatasetLocator(
                            "/%s/%s".format(namespace, name),
                            false,
                            request.mimeType())
                    DatasetConfig(
                            request.metadata,
                            locator,
                            request.extensions()
                    )
                } catch (e: IOException) {
                    throw RuntimeException("Exception while preprocessing file", e)
                }
            }
            else -> throw BadRequestException("Unknown request type '${request.javaClass.simpleName}'")
        }

        return catalogue.registerDataset(namespace, datasetConfig).id
    }

    private fun preprocessFile(namespace: String, fileName: String, fileType: FileType): String {
        val stream: InputStream = howl.downloadFile(namespace, fileName)
                ?: throw NotFoundException("File not found: " + fileName)

        return stream.use { s ->
            when (fileType) {
                FileType.GEOJSON -> {
                    try {
                        GeoJsonParser(s).validate()
                    } catch (e: Exception) {
                        throw BadRequestException("Exception while validating GeoJSON", e)
                    }
                    fileName
                }
                FileType.CSV -> {
                    val storageId = howl.uploadAnonymousFile(namespace, MediaType.APPLICATION_JSON) { outputStream ->
                        try {
                            CsvConverter().convert(s, outputStream)
                        } catch (e: IOException) {
                            throw BadRequestException("Exception while converting CSV to GeoJSON", e)
                        }
                    }
                    storageId.uid
                }
                FileType.RAW -> fileName
                else -> fileName
            }
        }
    }

    // No need for auth injection here, as there's no interaction with dataset namespaces
    @POST
    @Path("/file/{namespace}")
    @Produces(MediaType.APPLICATION_JSON)
    fun uploadFile(
            @PathParam("namespace") namespace: DatasetNamespace,
            @Context request: HttpServletRequest): HowlStorageId {
        return howl.uploadAnonymousFile(namespace.namespace, request.contentType) { outputStream ->
            try {
                IOUtils.copy(request.inputStream, outputStream)
            } catch (e: Exception) {
                throw RuntimeException("Exception while uploading file: " + e)
            }
        }
    }

    private fun notFoundException(type: String, name: String) = NotFoundException("$type '$name' not found")

    private fun throwIfDatasetNotPresentOrNotAllowed(user: User, coords: DatasetCoordinates) {
        val datasets = getDatasets(user)    // These will already be filtered to those that the user is authorised for

        val datasetsInNamespace = datasets[coords.namespace]
                ?: throw notFoundException("Namespace", coords.namespace.namespace)
        if (!datasetsInNamespace.contains(coords.id)) {
            throw notFoundException("Dataset", coords.id.uid)
        }
    }

    private fun throwIfNamespaceNotAllowed(user: User, namespace: DatasetNamespace) {
        if (!authoriser.authorisedFor(user, namespace)) {
            throw notFoundException("Namespace", namespace.namespace) // 404 instead of 403 to prevent discovery
        }
    }
}
